File: Iteration/229. Majority Element II.cpp
// 这道题的精华在于 C++ 里面 iterator 的erase操作。 it = m.erase(it)。代码19行到25行。

class Solution {
public:
    vector<int> majorityElement(vector<int>& nums)
    {
        int k = 2;
        unordered_map<int, int> m;
        for(int i=0; i < nums.size(); i++)
        {
            if(m.find(nums[i]) == m.end() && m.size() < k)
            {
                m[nums[i]] = 1;
            }
            else if(m.find(nums[i]) != m.end()){
                m[nums[i]] += 1;
            }
            else{
                for(auto it=m.begin(); it != m.end();)
                {
                    it->second--;
                    if(it->second == 0)
                        it = m.erase(it);
                    else
                        it++;
                }
            }
        }
        for(auto it=m.begin(); it != m.end(); it++)
        {
            it->second = 0;
        }
        
        vector<int> res;
        for(auto num:nums)
        {
            if(m.find(num) != m.end())
                m[num]++;
        }
        
        for(auto it=m.begin(); it != m.end(); it++)
        {
            if(it->second > nums.size() / (k + 1.0))
                res.push_back(it->first);
        }
        return res;
    }
};

File: Iteration/229. Majority Element II.py
"""
非常好的一道题。最关键的地方在于，如何iterate dictionary！
"""

class Solution:
    def majorityElement(self, nums: List[int]) -> List[int]:
        d = dict()
        for num in nums:
            if num in d:
                d[num] += 1
            elif len(d) < 2:
                d[num] = 1
            else:
                for key, value in list(d.items()):
                    d[key] -= 1
                    if d[key] == 0:
                        del d[key]
        for key in d:
            d[key] = 0
        for num in nums:
            if num in d:
                d[num] += 1
        res = []
        for key in d:
            if d[key] > len(nums) / 3:
                res.append(key)
        return res
